File: prep/implementation/patterns_reference_implementation/two_heaps/maximize_capital.cpp
#include "PriorityQueue.h"
#include <algorithm>
#include <iomanip>
#include <iostream>
#include <string>
#include <vector>

using namespace std;

// O(nlogn + klogn) / O(n)
int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, k, c, n;
  cin >> tt;

  while (tt--) {
    cin >> k >> c;
    cin >> n;

    vector<int> capital(n), profit(n);
    for (int i = 0; i < n; i++) {
      cin >> capital[i];
    }

    for (int i = 0; i < n; i++) {
      cin >> profit[i];
    }

    PriorityQueue<int> min_capital_heap(
        [&](int x, int y) { return capital[x] < capital[y]; });
    PriorityQueue<int> max_profit_heap(
        [&](int x, int y) { return profit[x] > profit[y]; });

    for (int i = 0; i < n; i++) {
      min_capital_heap.add(i);
    }

    int tot_cap = c;
    for (int i = 0; i < k; i++) {
      while (!min_capital_heap.empty() &&
             capital[min_capital_heap.peek()] <= tot_cap) {
        max_profit_heap.add(min_capital_heap.poll());
      }

      if (max_profit_heap.empty()) {
        break;
      }

      tot_cap += profit[max_profit_heap.poll()];
    }

    cout << tot_cap << "\n";
  }

  return 0;
}

File: lc/coreprep/implementation/practice_iterations/two_heaps_2/MaximizeCapital.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

// O(nlogn + klogn) / O(n)
public class MaximizeCapital {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int numProjects = in.nextInt();
        int initialCapital = in.nextInt();
        int n = in.nextInt();

        int[] capital = new int[n];
        for (int i = 0; i < n; i++) {
          capital[i] = in.nextInt();
        }

        int[] profit = new int[n];
        for (int i = 0; i < n; i++) {
          profit[i] = in.nextInt();
        }

        System.out.println(
            maximizeCapital(capital, profit, numProjects, initialCapital));
      }
    }
  }

  private static int maximizeCapital(int[] capital, int[] profit,
                                     int numProjects, int initialCapital) {
    PriorityQueue<Integer> minCapitalHeap =
        new PriorityQueue<>((x, y) -> Integer.compare(capital[x], capital[y]));
    PriorityQueue<Integer> maxProfitHeap =
        new PriorityQueue<>((x, y) -> Integer.compare(profit[y], profit[x]));

    int totalCapital = initialCapital;
    for (int i = 0; i < capital.length; i++) {
      minCapitalHeap.add(i);
    }

    for (int i = 0; i < numProjects; i++) {
      while (!minCapitalHeap.isEmpty() &&
             capital[minCapitalHeap.peek()] <= totalCapital) {
        maxProfitHeap.add(minCapitalHeap.poll());
      }

      if (maxProfitHeap.isEmpty()) {
        break;
      }

      totalCapital += profit[maxProfitHeap.poll()];
    }
    return totalCapital;
  }
}